contracts/ForwarderV4.sol

```diff
@@ -0,0 +1,380 @@
+  }
+
+  /**
+   * @inheritdoc IForwarderV4
+   */
+  function flushTokens(address tokenContractAddress)
+    external
+    virtual
+    override
+    onlyAllowedAddress
+  {
+    ERC20Interface instance = ERC20Interface(tokenContractAddress);
+    address forwarderAddress = address(this);
+    uint256 forwarderBalance = instance.balanceOf(forwarderAddress);
+    if (forwarderBalance == 0) {
+      return;
+    }
+
+    TransferHelper.safeTransfer(
+      tokenContractAddress,
+      parentAddress,
+      forwarderBalance
+    );
+  }
+
+  /**
+   * @inheritdoc IForwarderV4
+   */
+  function flushERC721Token(address tokenContractAddress, uint256 tokenId)
+    external
+    virtual
+    override
+    onlyAllowedAddress
+  {
+    IERC721 instance = IERC721(tokenContractAddress);
+    require(
+      instance.supportsInterface(type(IERC721).interfaceId),
+      'The tokenContractAddress does not support the ERC721 interface'
+    );
+
+    address ownerAddress = instance.ownerOf(tokenId);
+    instance.transferFrom(ownerAddress, parentAddress, tokenId);
+  }
+
+  /**
+   * @inheritdoc IForwarderV4
+   */
+  function flushERC1155Tokens(address tokenContractAddress, uint256 tokenId)
+    external
+    virtual
+    override
+    onlyAllowedAddress
+  {
+    IERC1155 instance = IERC1155(tokenContractAddress);
+    require(
+      instance.supportsInterface(type(IERC1155).interfaceId),
+      'The caller does not support the IERC1155 interface'
+    );
+
+    address forwarderAddress = address(this);
+    uint256 forwarderBalance = instance.balanceOf(forwarderAddress, tokenId);
+
+    instance.safeTransferFrom(
+      forwarderAddress,
+      parentAddress,
+      tokenId,
+      forwarderBalance,
+      ''
+    );
+  }
+
+  /**
+   * @inheritdoc IForwarderV4
+   */
+  function batchFlushERC1155Tokens(
+    address tokenContractAddress,
+    uint256[] calldata tokenIds
+  ) external virtual override onlyAllowedAddress {
+    IERC1155 instance = IERC1155(tokenContractAddress);
+    require(
+      instance.supportsInterface(type(IERC1155).interfaceId),
+      'The caller does not support the IERC1155 interface'
+    );
+
+    address forwarderAddress = address(this);
+    uint256 length = tokenIds.length;
+    uint256[] memory amounts = new uint256[](tokenIds.length);
+    for (uint256 i; i < length; i++) {
+      amounts[i] = instance.balanceOf(forwarderAddress, tokenIds[i]);
+    }
+
+    instance.safeBatchTransferFrom(
+      forwarderAddress,
+      parentAddress,
+      tokenIds,
+      amounts,
+      ''
+    );
+  }
+
+  /**
+   * @inheritdoc IForwarderV4
+   */
+  function batchFlushERC20Tokens(address[] calldata tokenContractAddresses)
+    external
+    virtual
+    override
+    onlyAllowedAddress
+  {
+    address forwarderAddress = address(this);
+    uint256 length = tokenContractAddresses.length;
+    for (uint256 i; i < length; i++) {
+      ERC20Interface instance = ERC20Interface(tokenContractAddresses[i]);
```

Comment:
What happens if `tokenContractAddresses[i]` is not an erc20 contract address
also incase we pass the same erc20 contract address twice, I think forwarderBalance check will go through for both cases
In this case, do we see partial failures or the whole tx fails?

Reply:
If we pass the same token address multiple times, then it doesn't seem to be causing any failures or something, it just seem to be performing the tx once. https://sepolia.arbiscan.io/tx/0x82a819eed6b4ee0b87d40893bd92faf99db2f731074d1f65c548966e1a62face

Also if the token address passed is not a erc20 contract address, then it seems to be erroring during the gas estimate itself, so shouldn't be an issue of partial or whole tx failure